Detailed installation instructions for linux/macos/windows (#593)
ansible/ansible-navigator: docs/installation.md

# Installing ansible-navigator with execution environment support

> [podman_url]: https://podman.io/

Reviewer: Traditionally, the detached link definitions live below the content. Either in the bottoms of the paragraphs where they are used or sections or accumulated at the bottom of the whole document. Could you adjust this accordingly?

```{contents}

> :local:

Reviewer: Side note: I wish we could just use the Furo theme. Then, we wouldn't need the local ToC at all.

```

## Linux

> [podman_install_url]: https://podman.io/getting-started/installation

Reviewer: It's quite uncommon to use such identifiers in MD detached link definitions. Just use normal text as it is used in the prose.

[docker_for_linux_url]: https://docs.docker.com/engine/install/

### Requirements

- Either [podman][podman_url] or [Docker for Linux][docker_for_linux_url]
- Internet access (during initial installation)

### Install the desired container engine for execution environment support

- Follow the [podman installation instructions][podman_install_url] for the appropriate distribution.
- Follow the [Docker for Linux installation instructions][docker_for_linux_url] for the appropriate distribution.

### Install ansible-navigator

1. Install the python package manager using the system package installer (e.g.):

>    ```

Reviewer: I'd prefer the code blocks to have explicit highlighting. We could go for `bash` or alternatively, we could go for `console`. But the latter is used for CLI recordings with a prompt like `$` prepended to the commands followed by their example outputs.

Reviewer: Also, it seems like the copybutton extension is able to cut the prompts off.

   sudo dnf install python3-pip
   ```

1. Install ansible-navigator:

   ```
   python3 -m pip install ansible-navigator --user
   ```

1. Add the installation path to the user shell initialization file (e.g.):

   ```
   echo 'export PATH=$HOME/.local/bin:$PATH' >> ~/.profile
   ```

1. Refresh the PATH (e.g.):

   ```
   source ~/.profile
   ```

1. Launch ansible-navigator:

   ```
   ansible-navigator
   ```

1. `ansible-navigator` triggers a one-time download of the demo execution-environment image.

## macOS

> [docker_for_mac_url]: https://hub.docker.com/editions/community/docker-ce-desktop-mac

Reviewer: Just call it
```suggestion
[Docker desktop for Mac]: https://hub.docker.com/editions/community/docker-ce-desktop-mac
```
and you won't need that `[docker_for_mac_url]` at all.

[mac_podman_issue_url]: https://github.com/containers/podman/issues/8016

### Requirements

- [Docker desktop for Mac][docker_for_mac_url]
- macOS command line developer tools
- Internet access (during initial installation)

### Install the desired container engine for execution environment support

- Follow the [Docker Desktop for Mac][docker_for_mac_url] installation instructions

```{note}

> There is no convenient way to use `ansible-navigator` with [podman][podman_url] on macOS.  Native source mounts from macOS through the [podman][podman_url] machine into the execution environment are not currently available.

Reviewer: Would you wrap this very long line?

   ```
   xcode-select install
   ```

1. Install ansible-navigator:

   ```
   pip3 install ansible-navigator --user
   ```

1. Add the installation path to the PATH:

   ```
   echo 'export PATH=$HOME/Library/Python/3.8/bin:$PATH' >> ~/.zprofile
   ```

1. Refresh the PATH:

   ```
   source ~/.zprofile
   ```

1. Launch ansible-navigator:

   ```
   ansible-navigator
   ```

1. `ansible-navigator` triggers a one-time download of the demo execution-environment image.

## Windows with WSL2

[docker_desktop_for_windows_url]: https://hub.docker.com/editions/community/docker-ce-desktop-windows
[manage_docker_non_root_user_url]: https://docs.docker.com/engine/install/linux-postinstall/
[kubic_project_url]: https://build.opensuse.org/package/show/devel:kubic:libcontainers:stable/podman
[ubuntu_url]: https://ubuntu.com/
[wsl_2_install_url]: https://docs.microsoft.com/en-us/windows/wsl/install-win10

### Requirements

- [Windows Subsystem for Linux 2][wsl_2_install_url]
- Either [podman][podman_url] or [Docker Desktop for Windows][docker_desktop_for_windows_url]
- Internet access (during initial installation)

> ### Setup Windows Subsystem for Linux 2] with Ubuntu

Reviewer 2: ```suggestion
### Setup Windows Subsystem for Linux 2 with Ubuntu
```

1. Install [Windows Subsystem for Linux 2][wsl_2_install_url].
1. Install the [Ubuntu][ubuntu_url] 20.04 LTS Linux distribution from the Microsoft store.
1. Open PowerShell and set the default WSL 2 distribution:

   ```
   wsl --set-default ubuntu
   ```

1. Launch the Ubuntu] virtual machine from the Windows menu and complete the initial set-up.
1. From the Ubuntu terminal, create the `/dev/mqueue` directory:

   ```
   sudo mkdir /dev/mqueue
   ```

### Install the desired container engine for execution environment support

- Installation instructions for [podman][podman_url] on Ubuntu 20.04 LTS.

   ```{note}
   The podman package is available in the official repositories for Ubuntu 20.10 and newer.
   Since interim releases of Ubuntu are not available on the Microsoft Store for WSL the
   [Kubic project][kubic_project_url] package can be used.
   ```

   1. Update the ubuntu package index:

      ```
      sudo apt update
      ```

>    1. Install system dependencies for podman]:

Reviewer: ```suggestion
   1. Install system dependencies for [podman]:
```

# Getting setup with execution environment support on:

## macOS

### Requirements

- Docker desktop for Mac
- macOS command line developer tools
- Internet access (during initial installation)

> ### Installation using podman as the container engine

Author: doh!

- Until native source mounts from macOS through the `podman` machine into the execution environment is available, there is not a convenient way to use `podman` on macOS with `ansible-navigator` [Related](https://github.com/containers/podman/issues/8016)
### Installation using Docker Desktop for Mac as the container engine

- Install [Docker Desktop for Mac](https://hub.docker.com/editions/community/docker-ce-desktop-mac)
- Open a terminal and enter `xcode-select install`, proceed with the command line developer tools installation if prompted
- Install ansible-navigator `pip3 install ansible-navigator --user`
- Add the installation path to the PATH (e.g. `echo 'export PATH=$HOME/Library/Python/3.8/bin:$PATH' >> ~/.zshrc`)
- Refresh the PATH (e.g. `source ~/.zshrc`)
- Launch ansible-navigator: `ansible-navigator`
- A one-time download of the demo execution-environment image will happen

## Windows with WSL2

### Requirements

- Windows Subsystem for Linux 2
- Either `podman` or Docker Desktop for Windows
- Internet access (during initial installation)

> #### Install podman as the container engine

Author: ty for catching that

   - Update using the new repository: `sudo apt-get update`
   - Install `podman`: `sudo apt-get install podman`
   - Create the `/dev/mqueue` directory: `sudo mkdir /dev/mqueue`
   - Continue to the `ansible-navigator` installation instructions below

### Install Docker Desktop for Windows as the container engine

- Install [WSL 2](https://docs.microsoft.com/en-us/windows/wsl/install-win10)
- Install the Ubuntu Linux distribution from the Microsoft store
- Open PowerShell and run `wsl --set-default ubuntu` to set the default WSL 2 distribution
- Install [Docker Desktop for Windows](https://hub.docker.com/editions/community/docker-ce-desktop-windows)
- Launch the Ubuntu virtual machine from the Windows menu and complete the initial Ubuntu set-up
- From the Ubuntu terminal:
   - Complete the [Manage Docker as a non-root user](https://docs.docker.com/engine/install/linux-postinstall/) steps
   - Update the ubuntu package index: `sudo apt update`
   - Install the python package manager: `sudo apt install python3-pip`

>    - Install ansible-navigator: `pip install ansible-navigator`

Author: got it

   - Add the installation path to the PATH: `source ~/.profile`
   - Launch ansible-navigator: `ansible-navigator`
   - A one-time download of the demo execution-environment image will happen

> ### Install ansible-navigator

Author: much better :)

> ### Install ansible-navigator

Reviewer: But you didn't actually change it.

- From the Ubuntu terminal"
   - Install the python package manager: `sudo apt install python3-pip`

>    - Install ansible-navigator: `pip install ansible-navigator`

Author: not sure exactly why that's necessary but happy to add

>    - Install ansible-navigator: `pip install ansible-navigator`

Reviewer: For consistency mostly.

   - Continue to the `ansible-navigator` installation instructions below

### Installation using `Docker Desktop for Windows` as the container engine

- Install [WSL 2](https://docs.microsoft.com/en-us/windows/wsl/install-win10)
- Install the Ubuntu Linux distribution from the Microsoft store
- Open PowerShell and run `wsl --set-default ubuntu` to set the default WSL 2 distribution
- Install [Docker Desktop for Windows](https://hub.docker.com/editions/community/docker-ce-desktop-windows)
- Launch the Ubuntu virtual machine from the Windows menu and complete the initial Ubuntu set-up
- From the Ubuntu terminal:
   - Complete the [Manage Docker as a non-root user](https://docs.docker.com/engine/install/linux-postinstall/) steps
   - Update the ubuntu package index: `sudo apt update`
   - Install the python package manager: `sudo apt install python3-pip`
   - Install ansible-navigator: `python3 -m pip install ansible-navigator`
   - Add the installation path to the PATH: `source ~/.profile`
   - Launch ansible-navigator: `ansible-navigator`
   - A one-time download of the demo execution-environment image will happen

### Install `ansible-navigator`

> - From the Ubuntu terminal"

Reviewer: This outer list has just one item and looks weird when rendered. Maybe change it to a paragraph-level sentence?

Author: ty again, good suggestion

# Getting setup with execution environment support on:

## macOS

### Requirements

- Docker desktop for Mac
- macOS command line developer tools
- Internet access (during initial installation)

### Installation using `podman` as the container engine

> - Until native source mounts from macOS through the `podman` machine into the execution environment is available, there is not a convenient way to use `podman` on macOS with `ansible-navigator` [Related](https://github.com/containers/podman/issues/8016)

Reviewer 2: ```suggestion
- There is no convenient way to use `ansible-navigator` with `podman` on macOS.  Native source mounts from macOS through the `podman` machine into the execution environment are not currently available. See this [related issue] (https://github.com/containers/podman/issues/8016) for details.
```

> ### Installation using `Docker Desktop for Mac` as the container engine

Reviewer 2: ```suggestion
### Installing on macOS with `Docker Desktop for Mac`

- Install [Docker Desktop for Mac](https://hub.docker.com/editions/community/docker-ce-desktop-mac)
- Open a terminal and enter `xcode-select install`, proceed with the command line developer tools installation if prompted
- Install ansible-navigator `pip3 install ansible-navigator --user`
- Add the installation path to the PATH (e.g. `echo 'export PATH=$HOME/Library/Python/3.8/bin:$PATH' >> ~/.zshrc`)
- Refresh the PATH (e.g. `source ~/.zshrc`)
- Launch ansible-navigator: `ansible-navigator`

> - A one-time download of the demo execution-environment image will happen

Reviewer 2: ```suggestion
- `ansible-navigator` triggers a one-time download of the demo execution-environment image.
```

Reviewer 2: I'm guessing this is what happens, but in general, you want to show this as a present tense action here and where it is repeated in other procedures below.

Another couple of general comments:
 - Would be nice to number these. Markdown can't do this other than manually saying 1. 2. etc instead of the current bullet list dashes. Since these are long procedures, it does help to have a number so someone can say "it didn't work on step 3' for example.
 - End each one with a period.

Reviewer: @samccann I'm not so sure about that. IIRC you can use `1.` for each line and MD will use the right numbers turning it into an ordered list. Plus MyST is much more powerful than CommonMark.

Reviewer 2: @webknjaz could be. I don't write in .md so it could also depend on read-the-docs where this is currently hosted.

Reviewer: No, it has nothing to do with RTD. Rendering is happening within Sphinx.

Author: @samccann good idea WRT line numbering, I _always_ forget the periods :)

## Windows with WSL2

### Requirements

- Windows Subsystem for Linux 2
- Either `podman` or Docker Desktop for Windows
- Internet access (during initial installation)

> ### Installation using `podman` as the container engine

Reviewer 2: ```suggestion
### Installing on Windows with `podman`
```

- Install the Ubuntu Linux distribution from the Microsoft store
- Open PowerShell and run `wsl --set-default ubuntu` to set the default WSL 2 distribution
- Launch the Ubuntu virtual machine from the Windows menu and complete the initial Ubuntu set-up
- From the Ubuntu terminal:
   - Update the ubuntu package index: `sudo apt update`
   - Install system dependencies for `podman`: `apt-get install curl wget gnupg2`
   - Source the Ubuntu release: `source /etc/os-release`
   - Add the `podman` repository: `sudo sh -c "echo 'deb http://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_${VERSION_ID}/ /' > /etc/apt/sources.list.d/devel:kubic:libcontainers:stable.list"`
   - Download the GPG key: `wget -nv https://download.opensuse.org/repositories/devel:kubic:libcontainers:stable/xUbuntu_${VERSION_ID}/Release.key -O- | sudo apt-key add -`
   - Update using the new repository: `sudo apt-get update`
   - Install `podman`: `sudo apt-get install podman`
   - Create the `/dev/mqueue` directory: `sudo mkdir /dev/mqueue`
   - Continue to the `ansible-navigator` installation instructions below

> ### Installation using `Docker Desktop for Windows` as the container engine

Reviewer 2: ```suggestion
### Installing on Windows with `Docker Desktop for Windows` 
```

- Install [WSL 2](https://docs.microsoft.com/en-us/windows/wsl/install-win10)
- Install the Ubuntu Linux distribution from the Microsoft store
- Open PowerShell and run `wsl --set-default ubuntu` to set the default WSL 2 distribution
- Install [Docker Desktop for Windows](https://hub.docker.com/editions/community/docker-ce-desktop-windows)
- Launch the Ubuntu virtual machine from the Windows menu and complete the initial Ubuntu set-up
- From the Ubuntu terminal:
   - Complete the [Manage Docker as a non-root user](https://docs.docker.com/engine/install/linux-postinstall/) steps

>    - Update the ubuntu package index: `sudo apt update`

Reviewer 2: ```suggestion
   - Update the Ubuntu package index: `sudo apt update`
```
